In [1]:
# ── Core ────────────────────────────────────────────────
import os
os.environ["HF_HUB_DISABLE_IMPLICIT_TOKEN_WARNING"] = "1"
import re
import json
import random
import warnings
from pathlib import Path

# ── Data ────────────────────────────────────────────────
import numpy as np
import pandas as pd

# ── Visualisation ───────────────────────────────────────
import matplotlib.pyplot as plt
import seaborn as sns

# ── Machine Learning utilities ──────────────────────────
from sklearn.model_selection import train_test_split, KFold
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
from sklearn.decomposition import PCA
from sklearn.manifold import TSNE

# ── Deep Learning ───────────────────────────────────────
import torch
import torch.nn as nn
import torch.nn.functional as F
from torch.utils.data import Dataset, DataLoader

# ── Transformers ────────────────────────────────────────
from transformers import (
    AutoTokenizer,
    AutoModel,
    AutoModelForSequenceClassification,
    get_linear_schedule_with_warmup,
)

# ── Configuration ───────────────────────────────────────
warnings.filterwarnings("ignore")

from transformers import logging as hf_logging
hf_logging.set_verbosity_error()
os.environ["HF_HUB_DISABLE_IMPLICIT_TOKEN_WARNING"] = "1"

SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
if torch.cuda.is_available():
    torch.cuda.manual_seed_all(SEED)

DEVICE = torch.device("cuda" if torch.cuda.is_available() else "cpu")

PALETTE = ["#FF6B9D", "#4ABCE8", "#F5A623"]
sns.set_theme(style="whitegrid", palette=PALETTE)

LANGUAGES = ["en", "es", "pt", "it"]


# The Grammar of Luxury

### A multilingual model of luxury sentiment across guest reviews

---

> *"Some people think luxury is the opposite of poverty. It is not. It is the opposite of vulgarity."*
>
> — attributed to Coco Chanel

---

**Author:** Avgustina Daskalova
**Course:** SoftUni — Deep Learning, July 2026
**Repository:** [github.com/Gusti4ka/The-Grammar-of-Luxury](https://github.com/Gusti4ka/The-Grammar-of-Luxury)

---

## Abstract

Chanel drew the line not between the rich and the poor, but between the discreet and the vulgar — and in doing so, she made luxury a matter of *register*. Of what is said, and what is left unsaid. Of tone rather than price.

Register is a linguistic property. And linguistic properties do not survive translation intact.

A Brazilian guest writes *aconchegante*; an Italian, *raffinato*; an Englishwoman, *understated*. Each word arrives carrying its own grammar of expectation — a private arithmetic of what refinement means and how one is permitted to speak of it. The words are not synonyms. They are dialects of the same discretion.

This project asks whether a machine can learn that grammar across borders.

It stands on two prior works. **Lost in Translation** established that semantic translation loss $\mathcal{L}(c)$ is *instrument-dependent*: monolingual embeddings collapsed under the weight of crossing languages, while multilingual ones held. **Arte Discreta** built a 74-term taxonomy of luxury across five aspects — but spoke only English, and asked a binary question of a dataset holding two positive answers. An honest null result; a hypothesis the data could not test.

*The Grammar of Luxury* answers both.

The taxonomy becomes the instrument rather than the subject: through distant supervision, its 74 terms generate a **continuous aspect-sentiment score**, dissolving the small-$n$ constraint that silenced its predecessor. Contextual multilingual embeddings — LaBSE and XLM-R, the instruments *Lost in Translation* vindicated — are then fine-tuned to read that score across **English, Spanish, Portuguese, and Italian**.

A second trained model closes the circuit. A fine-tuned voice, narrating each report in the language it was written in — text returning to speech, the unstructured become audible.

Two modalities. Four languages. One grammar.

---

## Table of Contents

1. **[Data — The Multilingual Corpus](#1-data--the-multilingual-corpus)**
   - 1.1 MAiDE-up (English, Spanish, Italian)
   - 1.2 Brazilian Hotel Reviews (Portuguese)
   - 1.3 Unified Corpus
2. **[The Taxonomy of Luxury](#2-the-taxonomy-of-luxury)** — porting Arte Discreta's 74 terms
3. **[Distant Supervision](#3-distant-supervision)** — from taxonomy to continuous luxury score
4. **[Exploratory Analysis](#4-exploratory-analysis)** — the corpus at a glance
5. **[The Text Model](#5-the-text-model)** — fine-tuning LaBSE / XLM-R
   - 5.1 Tokenisation & Datasets
   - 5.2 Architecture
   - 5.3 Training
   - 5.4 Evaluation
6. **[Interpretability](#6-interpretability)** — attention across languages
7. **[The Voice Model](#7-the-voice-model)** — narrating the report
8. **[Conclusions](#8-conclusions)**
9. **[References](#9-references)**

---

## 1. Data — The Multilingual Corpus

Two datasets feed this project, one per source-language group.

### 1.1 MAiDE-up (English, Spanish, Italian)

The **MAiDE-up** corpus (Ignat et al., 2024) collects 20,000 hotel reviews from Booking.com across ten languages, half genuine and half generated by GPT-4 for a deception-detection task. Three of its languages — **English, Spanish, Italian** — are ours.

Two decisions shape how we use it:

- **Real reviews only.** The corpus was built to distinguish human from machine writing. That is not our question. A model of *luxury* must learn from how real guests actually speak, so we discard the AI-generated half and keep only the human-written reviews (`source == 0`).

- **Reunited text.** Each review is stored split across two columns — `Upside_Review` and `Downside_Review`. We recombine them into a single passage, since luxury sentiment lives in the whole utterance, not in praise or complaint alone.

In [2]:
# ── Download MAiDE-up dataset (10 languages; real + AI reviews) ──
import urllib.request

DATA_DIR = Path("data")
DATA_DIR.mkdir(exist_ok=True)

MAIDE_URL = (
    "https://raw.githubusercontent.com/MichiganNLP/"
    "multilingual_reviews_deception/master/data/all_data.csv"
)
maide_path = DATA_DIR / "maide_up_all.csv"

if not maide_path.exists():
    print("Downloading MAiDE-up dataset…")
    urllib.request.urlretrieve(MAIDE_URL, maide_path)
    print(f"Saved to {maide_path}")
else:
    print(f"Already present: {maide_path}")
    

Already present: data\maide_up_all.csv


In [3]:
# ── Load MAiDE-up and inspect structure ──
maide = pd.read_csv(maide_path)

print(f"Shape: {maide.shape[0]:,} rows × {maide.shape[1]} columns\n")
print("Columns:", list(maide.columns), "\n")
maide.head(3)

Shape: 19,985 rows × 12 columns

Columns: ['Unnamed: 0', 'Review_Language', 'City Name', 'Hotel Name', 'Upside_Review', 'Downside_Review', 'Review_Score', 'Sentiment', 'source', 'Prompt_Language', 'na_up_review', 'na_down_review'] 



,Unnamed: 0,Review_Language,City Name,Hotel Name,Upside_Review,Downside_Review,Review_Score,Sentiment,source,Prompt_Language,na_up_review,na_down_review
0,0,Chinese,Ankara,迪万卡拉汉酒店,性价比超高的酒店，虽然价格比较贵，但是享受到的服务环境都还不错。,NaN,10.0,POS,0,NaN,False,True
1,1,Chinese,Ankara,迪万卡拉汉酒店,奥斯曼风格的建筑很有特色，房间装修精致，距离安卡拉城堡非常近，附近餐厅多，纪念品小商店多,不知道是不是因为夏季的原因，房间的空调不太给力，房间热；酒店员工会在窗外聊天，不考虑客人是不...,4.0,NEG,0,NaN,False,False
2,2,Chinese,Ankara,迪万卡拉汉酒店,非常喜歡酒店的環境氛圍，特別是通過地下通道到達博物館，這是一個驚喜（但入住時酒店服務人員并未...,酒店的空調令人傷心，第一天房間空調不起作用，次日上午出門時告知酒店經理，回來后依然沒有處理，...,5.0,NEG,0,NaN,False,False


In [4]:
# ── Filter: real reviews only, in EN / ES / IT ──
LANG_NAMES = {"English": "en", "Spanish": "es", "Italian": "it"}

# Keep real reviews (source == 0) in our three MAiDE-up languages
mask = (maide["source"] == 0) & (maide["Review_Language"].isin(LANG_NAMES))
maide_real = maide[mask].copy()

# Combine upside + downside into one review text
maide_real["text"] = (
    maide_real["Upside_Review"].fillna("").astype(str) + " "
    + maide_real["Downside_Review"].fillna("").astype(str)
).str.strip()

# Normalise language to short codes
maide_real["lang"] = maide_real["Review_Language"].map(LANG_NAMES)

# Keep only what we need
maide_clean = maide_real[["lang", "text", "Review_Score", "Sentiment"]].reset_index(drop=True)

print(f"Real EN/ES/IT reviews: {len(maide_clean):,}\n")
print(maide_clean["lang"].value_counts(), "\n")
maide_clean.head(3)

Real EN/ES/IT reviews: 3,000

lang
en    1000
it    1000
es    1000
Name: count, dtype: int64 



,lang,text,Review_Score,Sentiment
0,en,Staff. Serkan türk was a great help during our...,10.0,POS
1,en,Nicely furnished room and nicely decorated lob...,9.0,POS
2,en,"Very central, near 2 metro stations. Room was ...",9.0,POS


> **⚠ Manual step required — please read before running.**
>
> Unlike MAiDE-up, this corpus cannot be downloaded programmatically: Mendeley serves it behind a browser download button. To reproduce this notebook:
>
> 1. Visit **https://data.mendeley.com/datasets/2w3kvrg97m/1**
> 2. Download `tripadvisorBrazilianHotelREviews.zip` (98.8 MB, CC BY 4.0)
> 3. Place it in the `data/` folder of this repository
>
> The cells below then unzip and load it. Every other dataset in this project downloads automatically.

In [5]:
# ── Unzip the Brazilian Portuguese corpus ──
import zipfile

pt_zip = DATA_DIR / "tripadvisorBrazilianHotelREviews.zip"

with zipfile.ZipFile(pt_zip, "r") as z:
    print("Files inside the archive:")
    for name in z.namelist():
        print("  ", name)
    z.extractall(DATA_DIR)

print("\nExtracted.")

Files inside the archive:
   datesBrazil.txt
   gradesBrazil.txt
   titlesBrazil.txt
   commentsBrazilNormalized.txt

Extracted.


In [6]:
# ── Load the Brazilian Portuguese corpus (4 parallel text files) ──
def read_lines(filename):
    """Read a text file into a list of stripped lines."""
    with open(DATA_DIR / filename, encoding="utf-8") as f:
        return [line.strip() for line in f]

grades   = read_lines("gradesBrazil.txt")
titles   = read_lines("titlesBrazil.txt")
comments = read_lines("commentsBrazilNormalized.txt")
dates    = read_lines("datesBrazil.txt")

print(f"grades:   {len(grades):,}")
print(f"titles:   {len(titles):,}")
print(f"comments: {len(comments):,}")
print(f"dates:    {len(dates):,}")


grades:   730,069
titles:   730,069
comments: 730,069
dates:    730,069


In [7]:
# ── Build Portuguese DataFrame and subsample to match other languages ──
pt_full = pd.DataFrame({
    "grade": grades,
    "title": titles,
    "comment": comments,
    "date": dates,
})

print(f"Full Portuguese corpus: {len(pt_full):,} reviews")

# Combine title + comment into one review text (mirrors the MAiDE-up treatment)
pt_full["text"] = (pt_full["title"] + " " + pt_full["comment"]).str.strip()

# Drop empty or very short reviews
pt_full = pt_full[pt_full["text"].str.len() > 20]

# Subsample 1,000 to match EN / ES / IT
pt_clean = pt_full.sample(n=1000, random_state=SEED).reset_index(drop=True)
pt_clean["lang"] = "pt"

print(f"After cleaning:         {len(pt_full):,}")
print(f"Subsampled to:          {len(pt_clean):,}\n")
pt_clean[["lang", "text", "grade"]].head(3)


Full Portuguese corpus: 730,069 reviews
After cleaning:         730,066
Subsampled to:          1,000



,lang,text,grade
0,pt,"Quartos confortáveis mas had Para o preço, não...",2
1,pt,Sabe aquele velhinho que deram um tapa?? O lug...,3
2,pt,Aconchegante Situado muito próximo à praia dos...,4


### 1.2 Brazilian Hotel Reviews (Portuguese)

Portuguese is absent from MAiDE-up, so it arrives from a second source: the **Brazilian Portuguese Hotel's Reviews Corpus** (Souza, Oliveira & Moreira, 2018), 730,069 TripAdvisor reviews gathered from hotels across all 26 Brazilian state capitals and the Federal District.

The corpus is distributed as four parallel text files — grades, titles, comments, dates — one line per review. We recombine title and comment into a single passage, matching the treatment given to MAiDE-up's split reviews.

Two adjustments make it comparable to the others:

- **Subsampling.** 730,069 reviews would drown out the 1,000 available per language elsewhere. We draw a random sample of **1,000**, so no language dominates the model's attention.

- **Scale.** TripAdvisor grades reviews from 1 to 5; Booking.com, the source behind MAiDE-up, uses 1 to 10. These are normalised to a common scale in §1.3.

### 1.3 The Unified Corpus

Four languages, two sources, one corpus.

The two datasets arrive with different conventions, so unification requires two normalisations:

- **A common rating scale.** MAiDE-up carries Booking.com scores (1–10); the Brazilian corpus carries TripAdvisor circles (1–5). Both are mapped onto a shared 0–1 interval, preserving relative position while discarding the arbitrary units of each platform.

- **A common shape.** Each review is reduced to three fields: the language it was written in, its text, and its normalised rating. Everything else — hotel names, cities, dates, sentiment labels — falls away. What remains is what the model will read.

The result is a balanced quartet: 1,000 reviews in each of English, Spanish, Portuguese, and Italian. No language speaks louder than another.

In [8]:
# ── Unify: one corpus, four languages, common scale ──

# Normalise ratings to 0–1
maide_part = maide_clean[["lang", "text"]].copy()
maide_part["rating"] = maide_clean["Review_Score"] / 10.0        # Booking.com 1–10

pt_part = pt_clean[["lang", "text"]].copy()
pt_part["rating"] = pt_clean["grade"].astype(float) / 5.0        # TripAdvisor 1–5

# Combine
corpus = pd.concat([maide_part, pt_part], ignore_index=True)
corpus = corpus.sample(frac=1, random_state=SEED).reset_index(drop=True)  # shuffle

print(f"Unified corpus: {len(corpus):,} reviews\n")
print(corpus["lang"].value_counts(), "\n")
print(corpus["rating"].describe(), "\n")
corpus.head(5)

Unified corpus: 4,000 reviews

lang
en    1000
pt    1000
es    1000
it    1000
Name: count, dtype: int64 

count    4000.000000
mean        0.704000
std         0.261297
min         0.100000
25%         0.400000
50%         0.800000
75%         0.900000
max         1.000000
Name: rating, dtype: float64 



,lang,text,rating
0,en,Room was spacious. Had sofas and a table. Clea...,0.4
1,pt,"AGRADÁVEL SURPRESA!! Esse hotel é bem novo, co...",1.0
2,en,It is very close to the VFS and Shivaji stadiu...,0.4
3,pt,Para viagem de negócios! Para o viajante à neg...,0.8
4,es,"Habitaciones muy cómodas y muy completas, exce...",0.9


In [9]:
# An authored taxonomy of fine-hospitality aspects.
# Each aspect maps to a vocabulary of terms that signal it in review text.
# This is domain knowledge, not learned — the analytic heart of the project.

LUXURY_TAXONOMY = {
    "service": [
        "attentive", "impeccable", "gracious", "anticipate", "sommelier",
        "knowledgeable", "professional", "seamless", "courteous", "discreet",
        "personalized", "polished", "warm welcome", "host", "staff",
    ],
    "culinary_precision": [
        "exquisite", "delicate", "balanced", "refined", "technique",
        "perfectly cooked", "tasting menu", "seasonal", "innovative",
        "artistry", "presentation", "plating", "nuanced", "sublime", "complex",
    ],
    "ambience": [
        "elegant", "intimate", "atmosphere", "decor", "lighting", "ambiance",
        "romantic", "sophisticated", "tasteful", "serene", "luxurious",
        "stylish", "cozy", "stunning", "view",
    ],
    "ingredients": [
        "fresh", "quality", "local", "house-made", "artisanal", "organic",
        "sourced", "premium", "wagyu", "truffle", "foie gras", "caviar",
        "imported", "rare", "seasonal",
    ],
    "experience": [
        "memorable", "unforgettable", "special occasion", "celebration",
        "worth it", "exceptional", "remarkable", "once in a lifetime",
        "experience", "journey", "treat", "splurge", "destination", "perfect evening",
    ],
}

n_terms = sum(len(v) for v in LUXURY_TAXONOMY.values())
print(f"Taxonomy: {len(LUXURY_TAXONOMY)} aspects, {n_terms} terms total")
for aspect, terms in LUXURY_TAXONOMY.items():
    print(f"  {aspect}: {len(terms)} terms")

Taxonomy: 5 aspects, 74 terms total
  service: 15 terms
  culinary_precision: 15 terms
  ambience: 15 terms
  ingredients: 15 terms
  experience: 14 terms


## 2. The Taxonomy of Luxury

*Arte Discreta* built a lexicon: 74 terms across five aspects, naming the dimensions along which fine hospitality announces itself. Service is not ambience; the precision of a kitchen is not the theatre of a room. The taxonomy was authored, not learned — domain knowledge given to the model rather than extracted from it.

It is reproduced here unchanged. But it carries a flaw that its original project never had to confront: **every term is English**.

Applied to a Spanish review, it matches nothing. *Aconchegante* is not *cozy*; *raffinato* is not *refined* — not to a string comparison, at least. A lexicon in one language is blind in three others.

There are two ways to open its eyes. One is to translate all 74 terms by hand into Spanish, Portuguese, and Italian — accurate, laborious, and quietly self-defeating: if the translator solves the cross-lingual problem, there is nothing left for the model to demonstrate.

The other is to let the embeddings do it. If LaBSE places *cozy* and *aconchegante* near each other in vector space — as *Lost in Translation* suggests it should — then the taxonomy can cross languages on its own, and we can watch it happen.

This project takes the second path, with the first held in reserve as verification: the model proposes; a native-level reader of all four languages judges.

In [10]:
# ── Load LaBSE (multilingual sentence encoder) ──
MODEL_NAME = "sentence-transformers/LaBSE"

print("Loading LaBSE — first run downloads ~1.8 GB, please wait…")
tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)
model = AutoModel.from_pretrained(MODEL_NAME).to(DEVICE)
model.eval()

print(f"Loaded on: {DEVICE}")

Loading LaBSE — first run downloads ~1.8 GB, please wait…


Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

Loaded on: cpu


In [11]:
# ── Embed the taxonomy terms ──
def embed(texts, batch_size=32):
    """Return L2-normalised LaBSE embeddings for a list of texts."""
    all_vecs = []
    for i in range(0, len(texts), batch_size):
        batch = texts[i : i + batch_size]
        enc = tokenizer(batch, padding=True, truncation=True,
                        max_length=128, return_tensors="pt").to(DEVICE)
        with torch.no_grad():
            out = model(**enc)
        vecs = out.last_hidden_state[:, 0]      # CLS pooling
        vecs = F.normalize(vecs, p=2, dim=1)
        all_vecs.append(vecs.cpu())
    return torch.cat(all_vecs)

taxonomy_terms  = [t for terms in LUXURY_TAXONOMY.values() for t in terms]
taxonomy_aspect = [a for a, terms in LUXURY_TAXONOMY.items() for _ in terms]

term_vecs = embed(taxonomy_terms)

print(f"Embedded {len(taxonomy_terms)} terms → {tuple(term_vecs.shape)}")


Embedded 74 terms → (74, 768)


In [12]:
# ── Build a candidate vocabulary per language ──
from collections import Counter

def extract_vocab(texts, min_count=5, min_len=4):
    """Collect reasonably frequent word forms from a list of texts."""
    counter = Counter()
    for t in texts:
        words = re.findall(r"\b[^\W\d_]{%d,}\b" % min_len, str(t).lower())
        counter.update(words)
    return [w for w, c in counter.items() if c >= min_count]

vocab_by_lang = {}
for lang in LANGUAGES:
    texts = corpus.loc[corpus["lang"] == lang, "text"]
    vocab_by_lang[lang] = extract_vocab(texts)
    print(f"{lang}: {len(vocab_by_lang[lang]):,} candidate words")

en: 920 candidate words
es: 784 candidate words
pt: 1,304 candidate words
it: 880 candidate words


### 2.1 Crossing the Language Barrier

The taxonomy speaks English. The corpus speaks four languages. To bring them into contact, both must be expressed in the same terms — not words, but vectors.

LaBSE was trained so that a sentence and its translation land in nearly the same place in a 768-dimensional space. Meaning is preserved; language is not. A Portuguese word and an English one become directly comparable by the angle between their vectors.

So the procedure is simple in outline: embed every word the corpus uses, embed every term the taxonomy names, and for each term ask which words in each language stand closest to it. If the embeddings behave as *Lost in Translation* found they do, the answers should be recognisable — *cozy* should summon *aconchegante*, *acogedor*, *accogliente*.

Whether they are recognisable is the first real test of this project's premise.

In [15]:
# ── Embed the candidate vocabulary of each language ──
# ~3,900 words on CPU takes several minutes; [*] means the cell is still running.

vocab_vecs = {}

for lang in LANGUAGES:
    vocab_vecs[lang] = embed(vocab_by_lang[lang])
    print(f"{lang}: embedded {len(vocab_by_lang[lang]):,} words")

print("\nAll vocabularies embedded.")

en: embedded 920 words
es: embedded 784 words
pt: embedded 1,304 words
it: embedded 880 words

All vocabularies embedded.


In [16]:
# ── For each taxonomy term, find the nearest word in each language ──

def nearest_terms(term_idx, lang, top_k=3):
    """Return the top_k words in `lang` closest to a taxonomy term."""
    sims = vocab_vecs[lang] @ term_vecs[term_idx]        # cosine (vectors are normalised)
    top = sims.topk(top_k)
    return [(vocab_by_lang[lang][i], round(s.item(), 3))
            for i, s in zip(top.indices, top.values)]

# A first look: does the model recognise the words we expect it to?
probe = ["cozy", "elegant", "attentive", "exquisite", "memorable"]

for term in probe:
    i = taxonomy_terms.index(term)
    print(f"\n{term.upper()}  ({taxonomy_aspect[i]})")
    for lang in ["es", "pt", "it"]:
        matches = nearest_terms(i, lang)
        pretty = ", ".join(f"{w} ({s})" for w, s in matches)
        print(f"  {lang}: {pretty}")


COZY  (ambience)
  es: acogedor (0.951), confortable (0.94), agradable (0.921)
  pt: aconchegante (0.954), aconchegantes (0.948), confortável (0.915)
  it: accogliente (0.958), confortevole (0.926), comodissimo (0.924)

ELEGANT  (ambience)
  es: lindo (0.847), bonita (0.842), bonito (0.841)
  pt: elegante (0.988), charmoso (0.862), nobre (0.848)
  it: bellissima (0.843), accogliente (0.814), discreta (0.806)

ATTENTIVE  (service)
  es: atento (0.975), atentos (0.957), cuidado (0.895)
  pt: atenciosa (0.959), atencioso (0.929), atenciosos (0.904)
  it: attento (0.983), attenzione (0.885), cortese (0.816)

EXQUISITE  (culinary_precision)
  es: excelente (0.897), excelentes (0.877), genial (0.877)
  pt: excelente (0.897), elegante (0.885), excelentes (0.877)
  it: eccellente (0.887), eccezionale (0.869), bellissima (0.859)

MEMORABLE  (experience)
  es: inmejorable (0.762), impecable (0.719), increíble (0.716)
  pt: inesquecível (0.879), esquecer (0.823), lamentável (0.742)
  it: accogli

### 2.2 The Translator as Witness

The model has made its proposals. Whether they are *correct* is not something the model can tell us — cosine similarity measures proximity in a learned space, not fidelity of meaning. A number of 0.95 is a claim, not a verdict.

So the second path, held in reserve, is called upon now. Twenty-five terms are drawn at random from the taxonomy; for each, the model's single best proposal in Spanish, Portuguese, and Italian is recorded. Seventy-five judgements in all, made by a reader at home in each of these languages, against one question:

> *Would a hotel guest writing in this language use this word where an English guest would use the taxonomy term?*

Not *is it a dictionary equivalent* — a stricter and less useful test — but *does it do the same work in a review*. `Excelente` is not the dictionary's *exquisite*, yet it may carry the same weight in a Portuguese sentence. That judgement requires a person.

What the exercise yields is a number the model could not produce about itself: the proportion of its cross-lingual guesses that a native reader would let stand. And, more interesting than the number, a pattern in the failures.

In [17]:
# ── Draw a random sample of terms for expert validation ──

rng = np.random.default_rng(SEED)
sample_idx = rng.choice(len(taxonomy_terms), size=25, replace=False)

rows = []
for i in sample_idx:
    for lang in ["es", "pt", "it"]:
        word, sim = nearest_terms(i, lang, top_k=1)[0]
        rows.append({
            "aspect":     taxonomy_aspect[i],
            "term_en":    taxonomy_terms[i],
            "lang":       lang,
            "proposed":   word,
            "similarity": sim,
            "accepted":   "",          # ← to be filled in by the human judge
        })

validation = pd.DataFrame(rows)
print(f"{len(validation)} proposals to judge\n")
validation.head(12)

75 proposals to judge



,aspect,term_en,lang,proposed,similarity,accepted
0,ingredients,caviar,es,parís,0.771,
1,ingredients,caviar,pt,cofre,0.788,
2,ingredients,caviar,it,dove,0.801,
3,culinary_precision,plating,es,pared,0.760,
4,culinary_precision,plating,pt,cobertura,0.777,
5,culinary_precision,plating,it,moquette,0.785,
6,service,polished,es,limpiado,0.844,
7,service,polished,pt,arrumado,0.822,
8,service,polished,it,pulite,0.823,
9,experience,perfect evening,es,noche,0.746,


In [18]:
# ── Where does the method work, and where does it fail? ──

rows = []
for i, term in enumerate(taxonomy_terms):
    for lang in ["es", "pt", "it"]:
        word, sim = nearest_terms(i, lang, top_k=1)[0]
        rows.append({
            "aspect":     taxonomy_aspect[i],
            "term_en":    term,
            "n_words":    len(term.split()),
            "lang":       lang,
            "proposed":   word,
            "similarity": sim,
        })

matches = pd.DataFrame(rows)

print("Similarity by term length:")
print(matches.groupby("n_words")["similarity"].agg(["count", "mean", "max"]).round(3), "\n")

print("Similarity by aspect:")
print(matches.groupby("aspect")["similarity"].mean().sort_values(ascending=False).round(3), "\n")

print(f"Above 0.90: {(matches['similarity'] >= 0.90).sum()} of {len(matches)} "
      f"({100*(matches['similarity'] >= 0.90).mean():.0f}%)")

Similarity by term length:
         count   mean    max
n_words                     
1          198  0.864  1.000
2           21  0.795  0.944
4            3  0.693  0.706 

Similarity by aspect:
aspect
ambience              0.892
service               0.890
experience            0.849
ingredients           0.830
culinary_precision    0.816
Name: similarity, dtype: float64 

Above 0.90: 71 of 222 (32%)
